src/GitHub.Exports/Extensions/SimpleRepositoryModelExtensions.cs

```diff
@@ -30,5 +35,58 @@ public static bool MightContainSolution(this ISimpleRepositoryModel repository)
                       .Any(x => ((x.Attributes.HasFlag(FileAttributes.Directory) || x.Attributes.HasFlag(FileAttributes.Normal)) &&
                                 !x.Name.StartsWith(".", StringComparison.Ordinal) && !x.Name.StartsWith("readme", StringComparison.OrdinalIgnoreCase)));
         }
+
+        public static string CurrentSha(this ISimpleRepositoryModel repository)
+        {
+            var repo = GitService.GitServiceHelper.GetRepo(repository.LocalPath);
+            return repo.Commits.FirstOrDefault()?.Sha;
+        }
+
+        public static Uri GenerateUrl(this ISimpleRepositoryModel repository, string path = null,
+            int startLine = -1, int endLine = -1)
+        {
+            var cloneUrl = repository.CloneUrl;
+            if (cloneUrl == null)
+                return null;
+
+            if (path != null && Path.IsPathRooted(path))
+            {
+                // if the path root doesn't match the repository local path, then ignore it
+                if (repository.LocalPath == null ||
+                    !path.StartsWith(repository.LocalPath, StringComparison.OrdinalIgnoreCase))
+                {
+                    Debug.Assert(false, String.Format(CultureInfo.CurrentCulture, "GenerateUrl: path {0} doesn't match repository {1}", path, repository.LocalPath));
+                    path = null;
+                }
+                else
+                    path = path.Substring(repository.LocalPath.Length + 1);
+            }
+
+            var commitSha = repository.CurrentSha();
+            var uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
+            return new UriString(uri).ToUri();
```

Comment:
On line 66 you generate a URL using `cloneUrl.ToRepositoryUrl().AbsoluteUri` so you know this is going to be an `http` or `https` URL (that's what `ToRepositoryUrl` guarantees. So you _could_ simplify the previous two lines to...

```csharp
var uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
return new Uri(uri);
```

Alternatively, because we have an implicit conversion, you could also simplify it to...

```csharp
UriString uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
return new uri.ToUri();
```

I leave it to you. There's nothing explicitly wrong with what you have here. :smile: